### PR TITLE
neo4j: 3.5.14 -> 4.3.3

### PR DESCRIPTION
**Pull request by erictapen**
This is a rebase and continuation of https://github.com/NixOS/nixpkgs/pull/98443.

Running in the NixOS test the DB doesn't start up, so this is definetly not mergable yet.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/databases/neo4j.nix [hunk 1]
```diff
@@ -113,6 +117,8 @@ in {
     (mkRemovedOptionModule [ "services" "neo4j" "port" ] "Use services.neo4j.http.listenAddress instead.")
     (mkRemovedOptionModule [ "services" "neo4j" "boltPort" ] "Use services.neo4j.bolt.listenAddress instead.")
     (mkRemovedOptionModule [ "services" "neo4j" "httpsPort" ] "Use services.neo4j.https.listenAddress instead.")
+    (mkRemovedOptionModule [ "services" "neo4j" "shell" "enabled" ] "shell.enabled was removed upstream")
```
Comment (happysalada) [inline nixos/modules/services/databases/neo4j.nix:120]: should we also remove that option as well then ?

File: nixos/tests/neo4j.nix [hunk 2]
```diff
@@ -2,19 +2,25 @@ import ./make-test-python.nix {
...
+    server =
       { ... }:
 
       {
+	virtualisation.memorySize = 4096;
+	virtualisation.diskSize = 1024;
+
         services.neo4j.enable = true;
+        # require tls certs to be available
+        services.neo4j.https.enable = false;
```
Comment (happysalada) [inline nixos/tests/neo4j.nix:14]: I wonder why those default to true.
Most services i know about always expose only http by default. https is usually taken care of by a reverse proxy.

File: nixos/modules/services/databases/neo4j.nix [hunk 3]
```diff
@@ -35,25 +35,30 @@ let
...
+
+    # Directories (read and write)
+    dbms.directories.data=${cfg.directories.data}
+    dbms.directories.logs=${cfg.directories.home}/logs
+    dbms.directories.run=${cfg.directories.home}/run
 
     # HTTP Connector
     ${optionalString (cfg.http.enable) ''
-      dbms.connector.http.enabled=${boolToString cfg.http.enable}
+      dbms.connector.http.enabled=${boolToString cfg.https.enable}
```
Comment (happysalada) [inline nixos/modules/services/databases/neo4j.nix:59]: why do we change http to https, there is a separate https option right ?

File: nixos/doc/manual/release-notes/rl-2111.section.md [hunk 4]
```diff
@@ -311,3 +311,5 @@ To be able to access the web UI this port needs to be opened in the firewall.
 - Nginx will use the value of `sslTrustedCertificate` if provided for a virtual host, even if `enableACME` is set. This is useful for providers not using the same certificate to sign OCSP responses and server certificates.
 
 - `lib.formats.yaml`'s `generate` will not generate JSON anymore, but instead use more of the YAML-specific syntax.
+
+- Neo4j was updated from version 3 to version 4. See this [migration guide](https://neo4j.com/docs/upgrade-migration-guide/current/) on how to migrate your Neo4j instance.
```
Comment (bobby285271) [inline nixos/doc/manual/release-notes/rl-2111.section.md:315]: You should run [`./nixos/doc/manual/md-to-db.sh`](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/md-to-db.sh) to convert this to docbook, see https://nixos.org/manual/nixos/unstable/#chap-contributing.

File: nixos/tests/neo4j.nix [hunk 5]
```diff
@@ -2,19 +2,25 @@ import ./make-test-python.nix {
   name = "neo4j";
 
   nodes = {
-    master =
+    server =
       { ... }:
 
       {
+	virtualisation.memorySize = 4096;
+	virtualisation.diskSize = 1024;
```
Comment (bobby285271) [inline nixos/tests/neo4j.nix:10]: The CI suggested that you shouldn't use TABs here for indent.

**Comment by wegank** [comment]
Closed by #179710.

**Pull request closed without merging**